common/achievement/create.ts

```diff
@@ -0,0 +1,112 @@
+import { Prisma, achievement_type_enum } from '@prisma/client';
+import { Achievement_template, achievement_template_for_enum } from '../../graphql/generated';
+import { ActionID, SpecificNotificationContext } from '../notification/actions';
+import { prisma } from '../prisma';
+import { TemplateSelectEnum, getAchievementTemplates } from './template';
+import tracer from '../logger/tracing';
+import { AchievementToCheck } from './types';
+
+async function findUserAchievement<ID extends ActionID>(
+    templateId: number,
+    userId: string,
+    context: SpecificNotificationContext<ID>
+): Promise<AchievementToCheck> {
+    const keys = context ? Object.keys(context) : [];
+    const userAchievement = await prisma.user_achievement.findFirst({
+        where: {
+            templateId,
+            userId,
+            AND: keys.map((key) => {
+                return {
+                    context: {
+                        path: key,
+                        equals: context[key],
+                    },
+                };
+            }),
+        },
+        select: { id: true, userId: true, context: true, template: true, achievedAt: true, recordValue: true },
+    });
+    return userAchievement;
+}
+
+async function getOrCreateUserAchievement<ID extends ActionID>(
+    template: Achievement_template,
+    userId: string,
+    context: SpecificNotificationContext<ID>
+): Promise<AchievementToCheck> {
+    const isGlobal =
+        template.templateFor === achievement_template_for_enum.Global ||
+        template.templateFor === achievement_template_for_enum.Global_Courses ||
+        template.templateFor === achievement_template_for_enum.Global_Matches;
+    const existingUserAchievement: AchievementToCheck = await findUserAchievement(template.id, userId, !isGlobal ? context : undefined);
+    if (!existingUserAchievement) {
+        return await createAchievement(template, userId, context);
+    }
+    return existingUserAchievement;
+}
+
+const createAchievement = tracer.wrap('achievement.createAchievement', _createAchievement);
+async function _createAchievement<ID extends ActionID>(currentTemplate: Achievement_template, userId: string, context: SpecificNotificationContext<ID>) {
+    const templatesByGroup = await getAchievementTemplates(TemplateSelectEnum.BY_GROUP);
+    const templatesForGroup = templatesByGroup.get(currentTemplate.group).sort((a, b) => a.groupOrder - b.groupOrder);
+
+    const keys = Object.keys(context);
+    const userAchievementsByGroup = await prisma.user_achievement.findMany({
+        where: {
+            template: {
+                group: currentTemplate.group,
+            },
+            userId,
+            AND: keys.map((key) => {
```

Comment:
Do we actually need to search for the whole context here? Shouldn't it be enough to search by templateID + relation?